frontend/apps/reader/readerui.lua

```diff
@@ -593,28 +593,66 @@ end
 function ReaderUI:showReader(file, provider, seamless)
     logger.dbg("show reader ui")
 
-    file = ffiUtil.realpath(file)
     if lfs.attributes(file, "mode") ~= "file" then
         UIManager:show(InfoMessage:new{
              text = T(_("File '%1' does not exist."), BD.filepath(filemanagerutil.abbreviate(file)))
         })
         return
     end
 
-    if not DocumentRegistry:hasProvider(file) and provider == nil then
+    if provider == nil and DocumentRegistry:hasProvider(file) then
+        provider = DocumentRegistry:getProvider(file)
+    end
+    if provider ~= nil then
+        provider = self:extendProvider(file, provider)
+    end
+    if provider and provider.provider then
+        -- We can now signal the existing ReaderUI/FileManager instances that it's time to go bye-bye...
+        UIManager:broadcastEvent(Event:new("ShowingReader"))
+        self:showReaderCoroutine(file, provider, seamless)
+    else
         UIManager:show(InfoMessage:new{
             text = T(_("File '%1' is not supported."), BD.filepath(filemanagerutil.abbreviate(file)))
         })
         self:showFileManager(file)
-        return
     end
+end
 
-    -- We can now signal the existing ReaderUI/FileManager instances that it's time to go bye-bye...
-    UIManager:broadcastEvent(Event:new("ShowingReader"))
-    provider = provider or DocumentRegistry:getProvider(file)
-    if provider.provider then
-        self:showReaderCoroutine(file, provider, seamless)
+function ReaderUI:extendProvider(file, provider)
+    -- If file extension is pure "zip", check the archive content and choose the appropriate provider,
+    -- except when the provider choice is forced in the "Open with" dialog.
+    -- Also pass to crengine is_fb2 property, based on the archive content (pure "zip"),
+    -- or on the original file double extension ("fb2.zip" etc).
+    local _, file_type = filemanagerutil.splitFileNameType(file) -- supports double-extension
+    if file_type == "zip" then
+        -- read the content of zip-file and get extension of the 1st file
+        local std_out = io.popen("unzip -qql \"" .. file .. "\"")
```

Comment:
Aren't we back to "this might horribly blow up on Android" territory?

Reply:
yes

Reply:
It's the current approach, not changed.
Until we have a better unzipper.

Reply:
I guess the approach it isn't used at all on android. All modern android devices ship with an alternative `busybox` binary called `toybox`, which doesn't implement `unzip` yet.

Also AOSP don't enforce any particular set of commandline utilities, so on most devices you can expect that io.popen to fail.

If you have a test `zip` file that you can attach and its `std_out` I can try to do the same using JVM methods (like we do with tar files).

Reply:
Anyhow, not required for this PR.

Reply:
We already ship zlib btw, so something cross-platform could be preferable over Java.

Reply:
I disagree.

But mostly I'm not sure why we can't rely on `unzip` binary for platforms other than android.

Still, if we need something cross platform it would be best to rely on JVM things for android.

And sounds a bit to extreme to ship something in `koreader-base` and `android-luajit-launcher`. Anyhow the android part would require trivial modifications to make it compatible with the jni/java mess.

We did `unzip` for some time to extract assets. Same function could be used here.

Reply:
> But mostly I'm not sure why we can't rely on unzip 

@pazos 
Because local unzip sometimes sucks and doesn't support proper Cyrillic decoding, for example. See linked issues mentioned in the other branch here

Reply:
> And sounds a bit to extreme to ship something in koreader-base

But that's all I'm saying, we already do. :-) Zlib for extremely generic purposes, and additionally:
https://github.com/koreader/koreader-base/blob/60c0256494aa3fe244beab60fd579e0b2c8598e5/ffi/zipwriter.lua

Just pilfer a few other functions from the original source:
https://github.com/luarocks/luarocks/blob/eae4e554d00000025c7d03e8a17bdb7245c7beff/src/luarocks/tools/zip.lua

I don't think that'd be very different from doing it in Java.

Anyway, just pointing out the option's there.

Reply:
> Just pilfer a few other functions from the original source

I do not see functions to get the archive content, and to unzip selected file(s).

Reply:
We can repurpose zlib (adding it to jni dir in android-luajit-launcher) but it will still require some glue code to run under the JVM and there're no speed advantages at all over a plain java/c++ counterpart.

The rest of platforms are ok with zlib over ffi.

Reply:
> I do not see functions to get the archive content, and to unzip selected file(s) in the original source.

Alright, good to know.